### PR TITLE
Update readme, add CoC and Authors

### PR DESCRIPTION
**Pull request by Phrozyn**
# MozDef Event Framework

A PoC using AWS + Serverless framework

This is a work in progess which will eventually be used as a template for external event sources.

The idea behind this work was to create a repeatable process for introducing new event sources to MozDef without having to specially craft it every time. While this is new and we haven't onboarded many sources yet, it is expected the function library may grow, and with it we will introduce further changes to this framework as it becomes more mature. 

For example, being able to handle SSO, JWT, or other methods of authorizing event sources are in our interest to support, as well as the ability to reach out and pull down data from Rest API's using those methods of authorization. With the introduction of this framework we should be able to have a repeatable CI/CD pipeline, and be able to add to the library many functions to handle various types of connectivity and data retrieval without much additional work.
___
## Framework Components:

__Cloudformation:__ Cloudformation will create the CI/CD components utilizing:
- __CodePipeline__
- __CodeCommit__
- __CodeBuild__
- __s3__
- __CloudWatch__
- __GitHub__.

The intended behavior is to pull the framework from GitHub, and the custom configuration for a specific event source from its CodeCommit repository. The Cloudformation template has the potential to be separately deployed for various event sources, while the only difference will be the event source's CodeCommit repository that contains the custom configuration data for that event source. This allows us to maintain a generic framework that is event source agnostic and relieved of sensitive data. This structure presents the ability to maintain secret data, and preserve it's confidentiality.

__Serverless:__ The Serverless framework is utilized to deploy the actual application that will handle the event source data. This framework creates the following resources:
 - A custom authorizer for webhook based APIs that use a header token.
 - An API Gateway for receiving post events using the authorizer.
 - Custom configuration variables that are populated by the `buildspec.yml` located in each event source configuration's CodeCommit repository.
 - A lambda event handler that will validate to some degree the event to be processed by MozDef. At some point we may extend functionality here to fully process the event by another lambda.
 - An SQS queue to receive the events that have been handled by our lambda handler.
 - A Dead Letter Queue (DLQ) to receive events that fail to be handled, so that they can be reprocessed once a fix is introduced. This is future functionality that has not been implemented yet.

We plan to include testing modules in this framework as well.

To deploy this framework you will need the following established:

1. You'll need to design your parameter store pathing, and add your tokens or other data to be used with the framework. We used the following structure:

`/<Project_name>/<event_source_name>/<environemnt>/<authorizer_token_var>/`

This allows us to keep track of the various event sources, what environment they are used in, and keep the same variables across all event sources that will contain different values.

2. An AWS Account with the ability to use all of the AWS services previously mentioned.

3. A CodeCommit repository that contains your buildspec and deploy script.

Please note the env: variables that must be filled out, these can essentially be whatever you want them to be.
___

__Example of a Buildspec:__

```
version: 0.2 
 
env:  
  variables: 
    STAGE: dev 
    SERVICE: zoom 
    PROJECT: MozDef-EF 
    API_PATH: zoom 
  ssm: 
 
phases: 
  install: 
    runtime-versions: 
      python: 3.7 
      nodejs: 10 
    commands: 
      # Install dependencies here 
      - pip3 install --upgrade awscli -q 
      - npm install -g --silent --progress=false serverless 
      - npm install --silent --save-dev serverless-pseudo-parameters 
      - npm install --silent --save-dev serverless-prune-plugin 
  pre_build: 
    commands: 
      # Perform pre-build actions here, maybe tests for instance 
  build: 
    commands: 
      # Invoke the deploy script here 
      - chmod +x $CODEBUILD_SRC_DIR/config/deploy.sh 
      - $CODEBUILD_SRC_DIR/config/deploy.sh deploy $STAGE $AWS_REGION 
```

___    

**Example of the deploy script**

```
#!/bin/bash     
 
# This is deploy script template. Feel free modify per service/resource. 
   
instruction()   
{   
  echo "-----------------------------------------------"   
  echo "usage: ./deploy.sh deploy <env>"   
  echo "env: eg. dev, staging, prod, ..."   
  echo "for example: ./deploy.sh deploy dev"   
  echo ""   
  echo "to test: ./deploy.sh <int-test|acceptance-test>"   
  echo "for example: ./deploy.sh acceptance-test"   
}  
  
if [ $# -eq 0 ]; then 
  instruction   
  exit 1   
elif [ "$1" = "int-test" ] && [ $# -eq 1 ]; then 
  pytest int-test   
   
elif [ "$1" = "acceptance-test" ] && [ $# -eq 1 ]; then 
  pytest acceptance-test   
   
elif [ "$1" = "deploy" ] && [ $# -eq 3 ]; then 
  STAGE=$2   
  REGION=$3 
  STATE=`aws cloudformation describe-stacks --stack-name "$SERVICE-$STAGE" \ 
        --query Stacks[*].StackStatus --output text | grep -E "ROLLBACK|FAIL" -c` 
  # Forcefully remove the stack deployed by Serverless   
  # framework ONLY IF previous build errored  
  # NOTE: This is probably only a good idea for dev stage  
  if [ $STATE -ne 0 ]; then 
     sls remove -s $STAGE -r $REGION --force 
  fi   
  sleep 2  
  # Try to deploy again after stack removal  
  sls deploy -s $STAGE -r $REGION --force 
 
else   
  instruction   
  exit 1   
fi  
```
4. The GitHub Framework Repo which contains the initial Cloudformation template, this will automatically create your CI/CD pipeline and build out the serverless application as a part of that. You should not need to make any changes to it unless you plan to modify it for your own purposes, which is entirely OK for us!
5. An S3 bucket to hold the merger lambda code (merge.zip) and your zipped CodeCommit repository for your specific event source, mark this bucket private (not public), the CF template will create the CodeCommit Repo for you with your code. This will only need to be done once per event source, once the repo is created, you may commit to it via git or directly through aws-cli or the console.

## RUN IT
Upload the cloudformation template, fill in the CloudFormation parameters and environment variables according to your design, unless you intend to use the defaults, and wait for it to complete!
To run multiple deployments (e.g if you have multiple event sources you would like to consume), you'll need to rename the cloudformation template each time you upload it for a new deployment.

### CODE REVIEW

File: TODO.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+TODO
+----
+
+- Implement a separate lambda function to monitor the events queue and perform validation
```
Comment (caggle) [inline TODO.md]: I have added this, but feel free to remove it as in the readme we say "at some point we may do this".

File: README.md [hunk 2]
```diff
@@ -1,3 +1,134 @@
...
+
+__Cloudformation:__ Cloudformation will create the CI/CD components utilizing:
+- __CodePipeline__
+- __CodeCommit__
+- __CodeBuild__
+- __s3__
+- __CloudWatch__
+- __GitHub__.
+
+The intended behavior is to pull the framework from GitHub, and the custom configuration for a specific event source from its CodeCommit repository. The Cloudformation template has the potential to be separately deployed for various event sources, while the only difference will be the event source's CodeCommit repository that contains the custom configuration data for that event source. This allows us to maintain a generic framework that is event source agnostic and relieved of sensitive data. This structure presents the ability to maintain secret data, and preserve it's confidentiality.
```
Comment (caggle) [inline README.md]: Not sure if the last sentence of this paragraph is necessary, to me it kind of sounds surplus. But this is an opinion, feel free to disagree :)

File: README.md [hunk 3]
```diff
@@ -1,3 +1,134 @@
...
+
+__Serverless:__ The Serverless framework is utilized to deploy the actual application that will handle the event source data. This framework creates the following resources:
+ - A custom authorizer for webhook based APIs that use a header token.
```
Comment (caggle) [inline README.md]: Consider replacing "a header token" with "an Authorization header".

File: README.md [hunk 4]
```diff
@@ -1,3 +1,134 @@
...
+
+To deploy this framework you will need the following established:
+
+1. You'll need to design your parameter store pathing, and add your tokens or other data to be used with the framework. We used the following structure:
+
+`/<Project_name>/<event_source_name>/<environemnt>/<authorizer_token_var>/`
+
+This allows us to keep track of the various event sources, what environment they are used in, and keep the same variables across all event sources that will contain different values.
+
+2. An AWS Account with the ability to use all of the AWS services previously mentioned.
```
Comment (caggle) [inline README.md:41]: I wonder if we should elaborate here a bit more, such as: "For convenience, the build pipeline can be created using the CloudFormation template with an AWS account with administrative rights. This should be OK, as both Cloudformation and Serverless framework create dedicated IAM roles associated with the resources. You can fine tune permissions for these roles the Cloudformation template and/or serverless configuration file (`serverless.yml`)."

This reads a bit long, but something along these lines. Or maybe it's too much detail?

File: README.md [hunk 5]
```diff
@@ -1,3 +1,134 @@
...
+        - npm install --silent --save-dev serverless-prune-plugin 
+    pre_build: 
+        commands: 
+        # Perform pre-build actions here, maybe tests for instance 
+    build: 
+        commands: 
+        # Invoke the deploy script here 
+        - chmod +x $CODEBUILD_SRC_DIR/config/deploy.sh 
+        - $CODEBUILD_SRC_DIR/config/deploy.sh deploy $STAGE $AWS_REGION 
+```
```
Comment (caggle) [inline README.md]: The indentation appears to be off here, maybe it's just Markdown

File: README.md [hunk 6]
```diff
@@ -1,3 +1,134 @@
...
+
+___    
+
+**Example of the deploy script**
```
Comment (caggle) [inline README.md:84]: Same in here with the indentation.

File: README.md (hunk 2 above)
Comment (Phrozyn) [reply README.md]: I wondered that myself :) I can remove!

File: README.md (hunk 3 above)
Comment (Phrozyn) [reply README.md]: Thanks, that does sound better

File: README.md (hunk 4 above)
Comment (Phrozyn) [reply README.md:41]: You are spot on, I wanted to mention the IAM stuff, but wasn't sure how to word it, this makes sense.

File: TODO.md (hunk 1 above)
Comment (Phrozyn) [reply TODO.md]: Maybe reword it to say " Implement a separate lambda function to perform schema validation"
We won't need one to monitor it, but we should perhaps design the xray monitoring around this with alerting in some way.

File: README.md (hunk 5 above)
Comment (Phrozyn) [reply README.md]: Fixed!

File: README.md (hunk 6 above)
Comment (Phrozyn) [reply README.md:84]: Fixed!

File: TODO.md (hunk 1 above)
Comment (caggle) [reply TODO.md]: Makes sense. I will change in a subsequent commit.

File: README.md (hunk 3 above)
Comment (Phrozyn) [reply README.md]: Added!

File: README.md (hunk 2 above)
Comment (Phrozyn) [reply README.md]: Removed!

File: README.md (hunk 4 above)
Comment (Phrozyn) [reply README.md:41]: I think I'll add this to the bottom of the Readme under the RUN IT section.

Comment (Phrozyn) [reply README.md:41]: Done, please review!

**Comment by caggle** [comment]
I made another commit (last) to suggest some small changes around Markdown formatting. Otherwise I think this is good!

**Review by Phrozyn — COMMENTED** [review]
lol I can't approve I made the PR :)

Go ahead and approve and merge :)

**Review by caggle — APPROVED** [review]
